File: apps/chat/libs/chat-storage/src/main/java/com/example/chat/storage/adapter/FriendshipRepositoryAdapter.java
package com.example.chat.storage.adapter;

import com.example.chat.domain.friendship.Friendship;
import com.example.chat.domain.friendship.FriendshipId;
import com.example.chat.domain.friendship.FriendshipRepository;
import com.example.chat.domain.user.UserId;
import com.example.chat.storage.mapper.FriendshipMapper;
import com.example.chat.storage.repository.JpaFriendshipRepository;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * FriendshipRepository 구현체 (Adapter)
 *
 * Hexagonal Architecture의 Adapter
 */
@Repository
public class FriendshipRepositoryAdapter implements FriendshipRepository {
    private final JpaFriendshipRepository jpaRepository;
    private final FriendshipMapper mapper;

    public FriendshipRepositoryAdapter(JpaFriendshipRepository jpaRepository, FriendshipMapper mapper) {
        this.jpaRepository = jpaRepository;
        this.mapper = mapper;
    }

    @Override
    @Transactional
    public Friendship save(Friendship friendship) {
        var entity = mapper.toEntity(friendship);
        var saved = jpaRepository.save(entity);
        return mapper.toDomain(saved);
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<Friendship> findById(FriendshipId id) {
        return jpaRepository.findById(id.value()).map(mapper::toDomain);
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<Friendship> findByUserIdAndFriendId(UserId userId, UserId friendId) {
        return jpaRepository.findByUserIdAndFriendId(userId.value(), friendId.value())
                .map(mapper::toDomain);
    }

    @Override
    @Transactional(readOnly = true)
    public List<Friendship> findAcceptedFriendsByUserId(UserId userId) {
        return jpaRepository.findAcceptedFriendsByUserId(userId.value()).stream()
                .map(mapper::toDomain)
                .collect(Collectors.toList());
    }

    @Override
    @Transactional(readOnly = true)
    public List<Friendship> findPendingRequestsByFriendId(UserId friendId) {
        return jpaRepository.findPendingRequestsByFriendId(friendId.value()).stream()
                .map(mapper::toDomain)
                .collect(Collectors.toList());
    }

    @Override
    @Transactional(readOnly = true)
    public List<Friendship> findPendingRequestsByUserId(UserId userId) {
        return jpaRepository.findPendingRequestsByUserId(userId.value()).stream()
                .map(mapper::toDomain)
                .collect(Collectors.toList());
    }

    @Override
    @Transactional(readOnly = true)
    public List<Friendship> findBlockedByUserId(UserId userId) {
        return jpaRepository.findBlockedByUserId(userId.value()).stream()
                .map(mapper::toDomain)
                .collect(Collectors.toList());
    }

    @Override
    @Transactional(readOnly = true)
    public List<Friendship> findFavoritesByUserId(UserId userId) {
        return jpaRepository.findFavoritesByUserId(userId.value()).stream()
                .map(mapper::toDomain)
                .collect(Collectors.toList());
    }

    @Override
    @Transactional
    public void deleteById(FriendshipId id) {
        jpaRepository.deleteById(id.value());
    }

    @Override
    @Transactional(readOnly = true)
    public boolean existsMutualFriendship(UserId userId, UserId friendId) {
        return jpaRepository.existsMutualFriendship(userId.value(), friendId.value());
    }

    @Override
    @Transactional(readOnly = true)
    public List<Friendship> findAllByUserId(UserId userId) {
        return jpaRepository.findAllByUserId(userId.value()).stream()
                .map(mapper::toDomain)
                .collect(Collectors.toList());
    }
}

File: apps/chat/libs/chat-storage/src/main/kotlin/com/example/chat/storage/adapter/FriendshipRepositoryAdapter.kt
package com.example.chat.storage.adapter

import com.example.chat.domain.friendship.Friendship
import com.example.chat.domain.friendship.FriendshipId
import com.example.chat.domain.friendship.FriendshipRepository
import com.example.chat.domain.user.UserId
import com.example.chat.storage.mapper.FriendshipMapper
import com.example.chat.storage.repository.JpaFriendshipRepository
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

/**
 * FriendshipRepository 구현체 (Adapter)
 *
 * Hexagonal Architecture의 Adapter
 */
@Repository
class FriendshipRepositoryAdapter(
	private val jpaRepository: JpaFriendshipRepository,
	private val mapper: FriendshipMapper
) : FriendshipRepository {

	@Transactional
	override fun save(friendship: Friendship): Friendship {
		val entity = mapper.toEntity(friendship)
		val saved = jpaRepository.save(entity)
		return mapper.toDomain(saved)
	}

	@Transactional(readOnly = true)
	override fun findById(id: FriendshipId): Friendship? {
		return jpaRepository.findById(id.value)
			.orElse(null)
			?.let { mapper.toDomain(it) }
	}

	@Transactional(readOnly = true)
	override fun findByUserIdAndFriendId(userId: UserId, friendId: UserId): Friendship? {
		return jpaRepository.findByUserIdAndFriendId(userId.value, friendId.value)
			?.let { mapper.toDomain(it) }
	}

	@Transactional(readOnly = true)
	override fun findAcceptedFriendsByUserId(userId: UserId): List<Friendship> {
		return jpaRepository.findAcceptedFriendsByUserId(userId.value)
			.map { mapper.toDomain(it) }
	}

	@Transactional(readOnly = true)
	override fun findPendingRequestsByFriendId(friendId: UserId): List<Friendship> {
		return jpaRepository.findPendingRequestsByFriendId(friendId.value)
			.map { mapper.toDomain(it) }
	}

	@Transactional(readOnly = true)
	override fun findPendingRequestsByUserId(userId: UserId): List<Friendship> {
		return jpaRepository.findPendingRequestsByUserId(userId.value)
			.map { mapper.toDomain(it) }
	}

	@Transactional(readOnly = true)
	override fun findBlockedByUserId(userId: UserId): List<Friendship> {
		return jpaRepository.findBlockedByUserId(userId.value)
			.map { mapper.toDomain(it) }
	}

	@Transactional(readOnly = true)
	override fun findFavoritesByUserId(userId: UserId): List<Friendship> {
		return jpaRepository.findFavoritesByUserId(userId.value)
			.map { mapper.toDomain(it) }
	}

	@Transactional
	override fun deleteById(id: FriendshipId) {
		jpaRepository.deleteById(id.value)
	}

	@Transactional(readOnly = true)
	override fun existsMutualFriendship(userId: UserId, friendId: UserId): Boolean {
		return jpaRepository.existsMutualFriendship(userId.value, friendId.value)
	}

	@Transactional(readOnly = true)
	override fun findAllByUserId(userId: UserId): List<Friendship> {
		return jpaRepository.findAllByUserId(userId.value)
			.map { mapper.toDomain(it) }
	}
}
